### PR TITLE
Add explicit Cache-Control headers

### PR DESCRIPTION
**Pull request by thibaultdelor**
See https://github.com/rust-lang-nursery/rustup.rs/issues/1358

### CODE REVIEW

**Comment by thibaultdelor** [comment]
@alexcrichton S3 metadata are used by CloudFront to send the header.
If possible I would suggest to monitor the impact of that on the next nightly build.
One of my concern is that I am wondering if CloudFront will reread the files the S3 bucket for each request.  I would expect CloudFront to have tight integration with S3 but that's not what [Amazon's documentation](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/DownloadDistS3AndCustomOrigins.html) suggests.

File: promote-release/src/main.rs [hunk 1]
```diff
@@ -362,6 +362,10 @@ upload-addr = \"{}/{}\"
                 .arg("cp")
                 .arg("--recursive")
                 .arg("--only-show-errors")
+                .arg("--metadata-directive")
```
Comment (thibaultdelor) [inline promote-release/src/main.rs]: Archives, being versioned can be cached so we have to avoid copying over the Cache-control header and replace it.

**Comment by alexcrichton** [comment]
Thanks! To confirm these cache headers indicate that the contents should not be cached, right?

If so, I'd actually imagine that all archived content could be pretty heavily cached as it in theory never changes!

**Comment by thibaultdelor** [comment]
Exactly, that's why for archiving I set the Control-Cache to public, which basically means "Cache all the time for everyone".

**Comment by alexcrichton** [comment]
Oh! Sorry I misread that, looks great to me, thanks so much!

**Comment by thibaultdelor** [comment]
As a side note, I don't think S3/ Cloudfront is a good long term strategy and I think we should think about building an API. We could still use S3/Cloufront to store/send files but not to determine appropriate versions etc...
Several reasons for that:
 1) Right now theres some logic duplication where rustup needs to know about S3. An api could abstract that away, minimising how coupled things are and allowing to change backend, optimising cache, etc.
 2) Invalidating Cloudfront is known to be flakey. (I had bad experience myself and I have heard several time the same feedback). You can potentially ends up with really weird issues like a stable,toml being invalidate but not stable.toml.sha256, resulting in rustup thinking it got a corrupted download.
 3) You could a LOT of optimizations with an API that you can't do at the moment. For example you wouldn't need stable to be a folder, just a soft link, avoiding copying distributions; You could dynamically change what are the recommended default components; etc.

**Comment by alexcrichton** [comment]
@thibaultdelor heh I'd definitely believe it! We've mostly optimized for our own sanity (aka simplicity) historically rather than other aspects, but I can certainly see that shifting as Rust is much larger than when rustup was originally written!

**Pull request merged**